OneCommunityGlobal/HGNRest
# Ilya log tool backend

<img width="614" alt="tools-log-task" src="https://github.com/OneCommunityGlobal/HGNRest/assets/105448137/16acc7ce-56f8-4cfa-a79f-d702edde8291">


# Description
Adds functionality for the construction manager to navigate to [bmdashboard/tools/log](http://localhost:3000/bmdashboard/tools/log) on the front end, select a project, select Check In or Check Out and view a list of all tool type available for the selected action (example: Power Drill, Jackhammer, etc). The manager can then select the items of that type they want to check in or our based on their individual human-readable code, and click "submit". Checking out a tool item will make it unavailable for another check out until it is checked back in. The action will also update the "available" and "using" fields on each tool type and logRecord field on each item in the database.  

## Related PRS (if any):
This backend PR is related to the [#2296](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/pull/2296) frontend PR.
To test this backend PR you need to checkout the #XXX frontend PR.
…

## Main changes explained:
- New route created: "/api/bm/tools/log"
- bmLogTools function added to bmToolController.js. This function iterates over the list of tool types and updates "available" and "using" fields. It also iterated over the list of tool items and updated the "logRecord" field.

## How to test:
1. Check the current branch **ilya_log_tool_backend** on the back end. Make sure you are using Node version 14.17.6. Run **npm install** and **npm run build && npm start** to run the backend locally
2. Check out the **ilya_log_tool_frontend** branch on the frontend. Run commands **npm install** and **npm run start:local** to launch the front end application
3. Log in as any user
4. Go to [/bmdashboard](http://localhost:3000/bmdashboard) and log in again
5. Navigate to [/bmdashboard/tools/log](http://localhost:3000/bmdashboard/tools/log).
6. Select a project from the dropdown, select an action (default is "Check In")
7. On the table under "Tool/Equipment Number" select a tool item or multiple items based on the individual code 
8. Press "Submit". A green toast message should display for every tool item selected
9. The table should rerender and the codes selected earlier should no longer be available for the same action
10. OPTIONAL: to verify the changes in the database, open Compass or Robo3t, connect to hgnData_dev, select ** buildingInventoryTypes** collection, enter this query: {"__t":"tool_type","name":"Power Drill"} (change name based on the tool type you selected for check in/out) and press "Find". 
11. Fields "using" and "available" should have more or fewer items after the form was submitted, depending on the action. "Check In" should take an item out of the "using" array and add it to "available". "Check Out" should take and item out of the "available" array and add it to "using".  

## Screenshots or videos of change
https://github.com/OneCommunityGlobal/HGNRest/assets/105448137/a2463358-a348-48fc-b65e-2cb4c8b2de1d

s:
